### ISSUE TITLE
Connections and Reconnections Multiple times 

### ISSUE THREAD

**Issue by android-bash** [hivemq/hivemq-mqtt-client#419]
## Expected behavior
Once connected, only if there is network issues or something, client should get reconnected. 
## Actual behavior
Frequent connections. Every minute addConnectedListener listener is being called.

### Reproducer code
This is how i connect. 

     mClient = MqttClient.builder()
                .useMqttVersion3()
                .serverHost(BuildConfig.MQTTIPADDRESS)
                .automaticReconnectWithDefaultConfig()
                .serverPort(BuildConfig.MQTTPORTNO.toInt())
                .identifier(UUID.randomUUID().toString())
                .webSocketConfig(MqttWebSocketConfig.builder().serverPath("mqtt").build())
                .sslWithDefaultConfig()
                .addConnectedListener { context: MqttClientConnectedContext? ->
                    Log.d(TAG, "Connected Yay")
                    listeners?.forEach {
                        it.onConnect();
                    }
                }
                .addDisconnectedListener { context: MqttClientDisconnectedContext ->
                    Log.d(TAG, "Disconnected " + context.cause.message)
                    listeners?.forEach {
                        it.onDisconnect()
                    }
                }
                .buildAsync()

    mClient.connectWith()
                .keepAlive(120)
                .simpleAuth()
                .username(BuildConfig.MQTTUSERNAME)
                .password(BuildConfig.MQTTPASSWORD.toByteArray())
                .applySimpleAuth()
                .send()

## Details
- Affected HiveMQ MQTT Client version(s): 1.1.0

My Log : 

2020-05-14 19:44:41.244 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:45:46.281 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:46:50.375 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:47:54.650 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:48:59.194 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:50:03.604 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:51:07.446 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:52:11.199 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:53:15.503 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:54:19.933 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:55:24.320 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:56:29.144 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:57:34.049 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:58:38.481 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 19:59:42.371 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay
2020-05-14 20:00:46.477 15284-15475/com.wrkspot.employee.dev D/MQTT_MANAGER: Connected Yay

**Comment by SgtSilvio** [comment]
Hi @android-bash 
Can you please give more detail about your problem?
If you are observing frequent reconnections, please provide the causes for the disconnects.
Also, as it seems you are running on Android, can you provide more information about the Android environment: Is the Android device in sleep/low energy mode? Are you using an emulator?

Btw. there are new versions of the client: the newest is 1.2.0.

**Comment by rohit5ram** [comment]
@SgtSilvio I face the same issue on Android MINIX X9 4.4.2 /5.1.1. I have enabled the auto reconnect default and use cleanStart(true). I didn't set keepAlive.
I constantly get the logs printed every 2 secs and at times 3 mins
` Mqtt client connected`
`Mqtt client disconnected`
I am using client 1.1.4
On a side note I tried upgrading to 1.2.0. It stopped working from then all together. The connection was never established. I am not even sure if the control is exiting.

`Mqtt5AsyncClient client = MqttClient.builder()
                    .useMqttVersion5()
                    .identifier(deviceId)
                    .serverHost(appContext.getString(R.string.mqtt_wss_broker))
                    .serverPort(443)
                    .sslWithDefaultConfig()
                    .webSocketConfig(MqttWebSocketConfig.builder().serverPath("mqtt").build())
                    .automaticReconnectWithDefaultConfig()
                    .addConnectedListener(context -> {
                        setMqttClientConnected(true);
                        Log.i(Util.TAG, LOG_LABEL + " Mqtt client connected");
                        publishOnline(appContext);
                    })
                    .addDisconnectedListener(context -> {
                        setMqttClientConnected(false);
                        Log.e(Util.TAG, LOG_LABEL + " Mqtt client disconnected");
                    })
                    .buildAsync();
            GlobalData.instance().setMqttClient(client);
            client.connectWith()
                    .cleanStart(true)
                    .willPublish(Mqtt5Publish.builder()
                            .topic(formatTopic(appContext, appContext.getString(R.string.mqtt_last_will_topic)))
                            .payload("".getBytes())
                            .qos(MqttQos.AT_LEAST_ONCE)
                            .build())
                    .send()
                    .whenCompleteAsync(((connAck, throwable) -> {
                        if (throwable == null) subscribe(appContext);
                        else LogUtils.logException(new Exception(throwable), LOG_LABEL);
                    }));
        }
    }

    @SuppressLint("NewApi")
    private static void subscribe(Context appContext) {
        GlobalData.instance().getMqttClient()
                .subscribeWith()
                .topicFilter(formatTopic(appContext,
                        appContext.getString(R.string.mqtt_response_topic)))
                .qos(MqttQos.AT_LEAST_ONCE)
                .callback((mqtt5Publish) -> getSubscriber().onReceived(appContext, mqtt5Publish))
                .send()
                .whenCompleteAsync((mqtt5SubAck, throwableSub) -> {
                    if (throwableSub != null) {
                        Log.i(Util.TAG, LOG_LABEL + " MqttSubscriber not subscribed");
                        setMqttClientSubscribed(false);
                        LogUtils.logException(new Exception(throwableSub), LOG_LABEL);
                    } else {
                        Log.i(Util.TAG, LOG_LABEL + " MqttSubscriber subscribed");
                        setMqttClientSubscribed(true);
                    }
                });
    }`

project level build.gradle
`classpath 'gradle.plugin.com.github.sgtsilvio.gradle:android-retrofix:0.3.5'
`
gradle dependencies in library build.gradle
`  
    implementation "com.hivemq:hivemq-mqtt-client:1.2.0`
gradle dependencies in app build.gradle
`
implementation "net.sourceforge.streamsupport:android-retrostreams:1.7.1"
    implementation "net.sourceforge.streamsupport:android-retrofuture:1.7.1"
`

Can you please help solving this ?

**Comment by SgtSilvio** [comment]
@rohit5ram 
Since 1.2.0 all optional dependencies are modularized. As you are using the WebSocket transport, which is an optional feature, you need to declare the following dependencies:
```groovy
implementation "com.hivemq:hivemq-mqtt-client:1.2.0"
implementation "com.hivemq:hivemq-mqtt-client-websocket:1.2.0"
```
Then version 1.2.0 should work for you.

Documentation for all optional features:
https://hivemq.github.io/hivemq-mqtt-client/docs/installation/#optional-features

**Comment by rohit5ram** [comment]
@SgtSilvio It worked after including the library.Thank you. 

Even the connect/disconnect logs are gone and the connection is stable. Will keep you posted if I see the problem again

**Comment by rohit5ram** [comment]
@SgtSilvio The connection drops happening every ~2min. That problem is still not resolved

```
05-18 11:10:54.108 11710-11839 MqttUtils Mqtt client disconnected
05-18 11:10:56.768 11710-11839 MqttUtils Mqtt client connected
05-18 11:11:47.248 11710-11839 MqttUtils Mqtt client disconnected
05-18 11:11:49.598 11710-11839 MqttUtils Mqtt client connected
05-18 11:13:42.248 11710-11839 MqttUtils Mqtt client disconnected
05-18 11:13:45.008 11710-11839: MqttUtils Mqtt client connected
```

**Comment by SgtSilvio** [comment]
@rohit5ram @android-bash 
Sorry for the delayed response.
Can you please provide the source and cause of the disconnects.
For example add the following disconnected listener and paste the output here:
```java
        .addDisconnectedListener(context -> {
            System.out.println(context.getSource());
            context.getCause().printStackTrace();
        });
```

**Comment by rohit5ram** [comment]
@SgtSilvio These are the error logs
```
SERVER
W/System.err: com.hivemq.client.mqtt.exceptions.ConnectionClosedException: Server closed connection without DISCONNECT.
W/dalvikvm: getStackTrace() called but no trace available
```

**Comment by SgtSilvio** [comment]
@rohit5ram 

1. Which broker are you using (name and version)?
1. Can you take a look at the broker logs why it disconnects the client?

**Comment by rohit5ram** [comment]
@SgtSilvio we use vernemq 1.6.2
Looks like there is a related bug-fix in 1.9.2. Let me try using latest vernemq. 
`Fix bug causing idle websocket connections to be closed after 60 seconds` https://github.com/vernemq/vernemq/releases/tag/1.9.2
https://github.com/vernemq/vernemq/releases

Can you help me with this issue? 
https://github.com/hivemq/hivemq-mqtt-client/issues/423
Thank you

**Comment by SgtSilvio** [comment]
As there was no activity anymore, I will close this issue.
Don't hesitate to comment or open a new issue.
